SQL: integer parameter validation in string functions (#58923)

wuyinan0126 opened the pull request:
Related to #58923

In insert, locate, substring function, when argument `start` or `length` is greater than Integer.MAX_INT OR less then Integer.MIN_INT + 1 (note that `start` needs to minus 1), it causes overflow and leads to unexpected results.

Commit:
SQL: integer parameter validation in string functions (#58923)

In insert, locate, substring function, when argument `start` or `length` is greater than Integer.MAX_INT OR less then Integer.MIN_INT + 1 (note that `start` need to minus 1), it causes overflow and leads to unexpected results.

elasticmachine commented:
Pinging @elastic/es-ql (:Query Languages/SQL)

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/InsertFunctionProcessor.java:
```diff
@@ -67,8 +67,15 @@ public static Object doProcess(Object input, Object start, Object length, Object
+                    + ((Number) length).longValue() + "]");
+        }
+        
+        if ((((Number) start).longValue() > Integer.MAX_VALUE) || (((Number) start).longValue() - 1 < Integer.MIN_VALUE)) {
```
Please check the arguments in their order, so first the `start` then the `length`.

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/SubstringFunctionProcessor.java:
```diff
@@ -59,8 +59,15 @@ public static Object doProcess(Object input, Object start, Object length) {
+                    + ((Number) length).longValue() + "]");
+        }
+    
+        if ((((Number) start).longValue() > Integer.MAX_VALUE) || (((Number) start).longValue() - 1 < Integer.MIN_VALUE)) {
```
Same here, first `start` then `length`.

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/LocateFunctionProcessor.java:
```diff
@@ -59,6 +59,11 @@ public static Integer doProcess(Object pattern, Object input, Object start) {
             throw new SqlIllegalArgumentException("A number is required; received [{}]", start);
         }
+    
+        if (start != null && ((((Number) start).longValue() > Integer.MAX_VALUE) || (((Number) start).longValue() - 1 < Integer.MIN_VALUE))) {
```
I'd have an `if (start != null)` as an upper level if and then have inside it a check like that:
```
if (start instanceof Number == false) {
...
} else if (<range check>) {
...
}
```

matriv reviewed (commented):
@wuyinan0126 Thank you for continuing on this.
I left some comments for the checks.
Additionally, Please extract the checks into methods (one for > 0, one for > MIN_VALUE) so they can be reused and probably pass the name of the argument as an arg to that method to have a relevant error message. Moreover, this way, you can avoid the duplication of cast to Number as this can be done when calling that new method.

Please also add unit tests to the corresponding ProcessorTests classes.

Commit:
SQL: IN operator incorrectly handling specified date format (#58932)

Extract the checks into methods Check.isNumberOutOfRange and add unit tests to the three string function.

wuyinan0126 commented:
@matriv Thanks for your advice and I have extracted the checks into methods Check.isNumberOutOfRange and added unit tests to the three string function.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/util/Check.java:
```diff
@@ -36,4 +36,14 @@ public static void notNull(Object object, String message, Object... values) {
         }
     }
+    
+    public static void isNumberOutOfRange(Object object, String objectName, Long from, Long to) {
```
The `Check` class' semantics seem to hint a naming in the lines of "check that <predicate>" (Check#isNull, Check#isTrue). Also, the function doesn't handle floating point `Number`s.
```suggestion
    public static void isFixedNumberAndInRange(Object object, String objectName, Long from, Long to) {
```

wuyinan0126 replied:
Thanks for the advice.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/util/Check.java:
```diff
@@ -36,4 +36,14 @@ public static void notNull(Object object, String message, Object... values) {
     }
+    
+    public static void isNumberOutOfRange(Object object, String objectName, Long from, Long to) {
+        if (!(object instanceof Number)) {
```
```suggestion
        if ((object instanceof Number) == false) {
```
This is a syntax part of [CONTRIBUTING.md](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md). Not everybody loves it, but it is more obvious.

wuyinan0126 replied:
Thanks for the advice.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/util/Check.java:
```diff
@@ -36,4 +36,14 @@ public static void notNull(Object object, String message, Object... values) {
+    
+    public static void isNumberOutOfRange(Object object, String objectName, Long from, Long to) {
+        if (!(object instanceof Number)) {
+            throw new SqlIllegalArgumentException("A number is required; received [{}]", object);
```
Since the object name is available, we might as well use it? We also want to log object's class name, not value.
```suggestion
            throw new SqlIllegalArgumentException("A number is required for[{}]; received [{}]", objectName, object.getClass());
```

wuyinan0126 replied:
Thanks for the advice.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/util/Check.java:
```diff
@@ -36,4 +36,14 @@ public static void notNull(Object object, String message, Object... values) {
+            throw new SqlIllegalArgumentException("A number is required; received [{}]", object);
+        }
+        if (((Number) object).longValue() > to || ((Number) object).longValue() < from) {
+            throw new SqlIllegalArgumentException("[{}] must be in the interval [{}..{}], but was [{}]", 
```
Optional:
```suggestion
            throw new SqlIllegalArgumentException("[{}] has the value [{}] out of allowed range [{}..{}]", 
```
or similar (and respective params swapped).

wuyinan0126 replied:
Thanks for the advice.

Commit:
SQL: IN operator incorrectly handling specified date format (#58932)

Update the Check.isFixedNumberAndInRange function to comply with code specifications.

Commit:
SQL: IN operator incorrectly handling specified date format (#58932)

Update the unit tests to fit new check error messages.

bpintea commented:
@wuyinan0126 , I was trying some other [string functions](https://www.elastic.co/guide/en/elasticsearch/reference/current/sql-functions-string.html), like `REPEAT` and it also exhibits the integer overflow behavior (as well as at lest the rest of functions in `BinaryStringNumericOperation` enum). There might be some more checks useful (such as if the argument to `StringBuffer`s would exceed their range) as input validation, although that's not necessarily the object of the initial ticket.
Would you care to expand the fix?

wuyinan0126 commented:
@bpintea OK, I will try it when I am free.

Commit:
SQL: integer parameter validation in string functions (#58923)

use Integer.MIN/MAX_VALUE to make it more clear in tests

bpintea commented:
@wuyinan0126 thanks for the fix, I've ported it to the new #63338 PR which I've opened to continue this PR (since we couldn't know if you'd still want to continue the work). That new PR also addresses the [comment above](https://github.com/elastic/elasticsearch/pull/60281#issuecomment-668178868), so I'd like to finish the work there, so I'll close this PR.
As you can see, you are mentioned as co-authoring the new PR and your commits have been kept, so proper credit is provided. Thank you again for your contribution!
